beer-garden/functionary
# #114 - Added backend support for scheduled tasks

Introduces backend support for Scheduled Tasks as part of  #114 

## Introduced Changes
- Removed `scheduler` app and its references from Functionary
- Added `ScheduledTask` model with several helper methods
  - Methods for updating the state of a `ScheduledTask` (`Active`, `Paused`, `Error`, `Archived`)
  - Method for updating the `last_run_at` field to keep track of the last time the task was run
- Updated `Task` model to include an optional reference to a `ScheduledTask`
- Added new celery task for executing a `ScheduledTask`
- Added new `functionary-scheduler` container that runs `celery-beat`

## How it works
The `ScheduledTask` essentially just holds a reference to a `PeriodicTask`, which uses the fields associated with the `ScheduledTask` in order to construct a new Task when it is time to run the celery task again.

The process for creating a `ScheduledTask` is as follows:
1. Create a `ScheduledTask` object by providing a `name`, `function`, and `parameters`
2. Create a `CrontabSchedule`
3. Create a `PeriodicTask`
    - Supply the `CrontabSchedule` to the `PeriodicTask`
    - The `PeriodicTask` should execute the `core.utils.tasking.run_scheduled_task` celery task
    - Supply the `ScheduledTask` UUID as a positional argument for the `PeriodicTask`
4. Add the `PeriodicTask` to the `ScheduledTask`
5. Enable the `PeriodicTask`

## Testing
- Launch Functionary and apply the migrations
- Login to the Admin UI
- Create a `ScheduledTask`
  - Leave the Periodic task field empty, since it will be assigned after we create the `PeriodicTask`
  - Leave status as `Pending`
  - Supply parameters for the function to be executed (can be copied from a previous task)
- Create a `Crontab` schedule
- Create a `PeriodicTask`
  - Use the `core.utils.tasking.run_scheduled_task` celery task
  - Use the `Crontab` created in the previous step
  - Provide the `ScheduledTask` UUID you created as a positional argument
  - You can enable the periodic task for the sake of testing. Otherwise, it is expected to start in the `disabled` state, and then get enabled after it has been assigned to a `ScheduledTask`
- Go back to the `ScheduledTask` and assign the newly created `PeriodicTask` to the Periodic task field
- View the `functionary-scheduler` container logs to see when `celery-beat` schedules the task
- View the `functionary-worker` container to see the tasks get executed
- Shell into the `functionary-django` container and start a django shell
  - `from core.models import ScheduledTask`
  - Get the `ScheduledTask` object you created
  - Enable the `PeriodicTask` by running the `.activate()` method on the `ScheduledTask`
  - Disable the `PeriodicTask` by running either `.pause()`, `.archive()`, or `.error()` methods on the `ScheduledTask`
    - Verify the `status` of the `PeriodicTask` is expected (PAUSED, ARCHIVED, ERROR, ACTIVE)
    - Verify the task is no longer getting scheduled
